mne/tests/test_annotations.py

```diff
@@ -1415,3 +1415,186 @@ def test_annotation_duration_setting():
+    raw.set_meas_date(_ORIG_TIME)
+    annotations = mne.Annotations(onset=[0, 19.990],
+                                  duration=[1, 1],
+                                  description=['test'] * 2)
+    print(raw)
+    with pytest.raises(RuntimeWarning) as e:
+        raw.set_annotations(annotations)
+        assert "expanding outside the data range" in str(e)
+
+
+def test_cropping_annotations_include_tmax():
+    """Test Raw cropping anntations including tmax"""
+    n_channels = 1
+    sampling_freq = 100
+    length_seconds = 30
+
+    info = mne.create_info(n_channels, sfreq=sampling_freq,
+                           ch_types=['eeg'])
+    data = np.random.randn(n_channels, sampling_freq * length_seconds)
+    raw = mne.io.RawArray(data, info)
+    annotations = mne.Annotations(onset=[0, 10, 20],
+                                  duration=[0, 10, 1],
+                                  description=['test'] * 3)
+    raw.set_annotations(annotations)
+    cropped = raw.copy().crop(tmin=0, tmax=20, include_tmax=True)
+    assert_equal(len(cropped.annotations), 3)
+
+
+def test_cropping_annotations_not_include_tmax():
+    """Test Raw cropping anntations including tmax"""
+    n_channels = 1
+    sampling_freq = 100
+    length_seconds = 20
+    info = mne.create_info(n_channels, sfreq=sampling_freq,
+                           ch_types=['eeg'])
+    data = np.random.randn(n_channels, sampling_freq * length_seconds)
+    raw = mne.io.RawArray(data, info)
+    annotations = mne.Annotations(onset=[0, 10, 19.990],
+                                  duration=[0, 2, 0],
+                                  description=['test'] * 3)
+    raw.set_annotations(annotations)
+    cropped = raw.copy().crop(tmin=0, tmax=19.99, include_tmax=False)
+    assert_equal(len(cropped.annotations), 2)
+
+
+def test_cropping_annotations_exceeds_data_range():
+    """Test Raw cropping anntations exceed """
+    n_channels = 1
+    sampling_freq = 100
+    length_seconds = 20
+    info = mne.create_info(n_channels, sfreq=sampling_freq,
+                           ch_types=['eeg'])
+    data = np.random.randn(n_channels, sampling_freq * length_seconds)
+    raw = mne.io.RawArray(data, info)
+    annotations = mne.Annotations(onset=[0, 10, 19.990],
+                                  duration=[0, 2, 0],
+                                  description=['test'] * 3)
+    raw.set_annotations(annotations)
+    with pytest.raises(ValueError) as e:
+        _ = raw.copy().crop(tmin=0, tmax=20, include_tmax=False)
+        assert "must be less than or equal to" in str(e)
+
+
+def test_cropping_annotations_on_boundaries():
+    """Test Raw cropping anntations on boundaries"""
+    n_channels = 1
+    sampling_freq = 100
+    length_seconds = 20
+
+    info = mne.create_info(n_channels, sfreq=sampling_freq,
+                           ch_types=['eeg'])
+    data = np.random.randn(n_channels, sampling_freq * length_seconds)
+    raw = mne.io.RawArray(data, info)
```

Comment:
@jevenzh you could use a pytest fixture to avoid recomputing such a raw toy data for each test. Tell us if you need help

Reply:
Oops, I indeed violated the DRY principle, thanks for the hint! I will learn about the pytest fixture and apply it here.